subsys/bluetooth/host/classic/l2cap_br.c

```diff
@@ -293,1712 +405,5198 @@ static void cancel_data_ready(struct bt_l2cap_br_chan *br_chan)
+	 * channel (service) security requirements against link security and
+	 * bt_conn_set_security API returns 0 what implies also there was no
+	 * need to trigger authentication.
+	 */
+	if (check == 0 && chan->conn->sec_level >= br_chan->required_sec_level) {
+		return L2CAP_CONN_SECURITY_PASSED;
+	}
+
+	/*
+	 * If 'check' still holds 0, it means local host just sent HCI
+	 * authentication command to start procedure to increase link security
+	 * since service/profile requires that.
+	 */
+	if (check == 0) {
+		/*
+		 * General Bonding refers to the process of performing bonding
+		 * during connection setup or channel establishment procedures
+		 * as a precursor to accessing a service.
+		 * For current case, it is dedicated bonding.
+		 */
+		atomic_set_bit(chan->conn->flags, BT_CONN_BR_GENERAL_BONDING);
+		return L2CAP_CONN_SECURITY_PENDING;
+	}
+
+	/*
+	 * For any other values in 'check' it means there was internal
+	 * validation condition forbidding to start authentication at this
+	 * moment.
+	 */
+	return L2CAP_CONN_SECURITY_REJECT;
+}
+
+static void l2cap_br_send_conn_rsp(struct bt_conn *conn, uint16_t scid, uint16_t dcid,
+				   uint8_t ident, uint16_t result)
+{
+	struct net_buf *buf;
+	struct bt_l2cap_conn_rsp *rsp;
+	struct bt_l2cap_sig_hdr *hdr;
+
+	buf = bt_l2cap_create_pdu(&br_sig_pool, 0);
+
+	hdr = net_buf_add(buf, sizeof(*hdr));
+	hdr->code = BT_L2CAP_CONN_RSP;
+	hdr->ident = ident;
+	hdr->len = sys_cpu_to_le16(sizeof(*rsp));
+
+	rsp = net_buf_add(buf, sizeof(*rsp));
+	rsp->dcid = sys_cpu_to_le16(dcid);
+	rsp->scid = sys_cpu_to_le16(scid);
+	rsp->result = sys_cpu_to_le16(result);
+
+	if (result == BT_L2CAP_BR_PENDING) {
+		rsp->status = sys_cpu_to_le16(BT_L2CAP_CS_AUTHEN_PEND);
+	} else {
+		rsp->status = sys_cpu_to_le16(BT_L2CAP_CS_NO_INFO);
+	}
+
+	l2cap_send(conn, BT_L2CAP_CID_BR_SIG, buf);
+}
+
+static int l2cap_br_conn_req_reply(struct bt_l2cap_chan *chan, uint16_t result)
+{
+	/* Send response to connection request only when in acceptor role */
+	if (!atomic_test_bit(BR_CHAN(chan)->flags, L2CAP_FLAG_CONN_ACCEPTOR)) {
+		return -ESRCH;
+	}
+
+	l2cap_br_send_conn_rsp(chan->conn, BR_CHAN(chan)->tx.cid, BR_CHAN(chan)->rx.cid,
+			       BR_CHAN(chan)->ident, result);
+	BR_CHAN(chan)->ident = 0U;
+
+	return 0;
+}
+
+#if defined(CONFIG_BT_L2CAP_DYNAMIC_CHANNEL)
+#if defined(CONFIG_BT_L2CAP_LOG_LEVEL_DBG)
+void bt_l2cap_br_chan_set_state_debug(struct bt_l2cap_chan *chan, bt_l2cap_chan_state_t state,
+				      const char *func, int line)
+{
+	struct bt_l2cap_br_chan *br_chan;
+
+	br_chan = BR_CHAN(chan);
+
+	LOG_DBG("chan %p psm 0x%04x %s -> %s", chan, br_chan->psm,
+		bt_l2cap_chan_state_str(br_chan->state), bt_l2cap_chan_state_str(state));
+
+	/* check transitions validness */
+	switch (state) {
+	case BT_L2CAP_DISCONNECTED:
+		/* regardless of old state always allows this state */
+		break;
+	case BT_L2CAP_CONNECTING:
+		if (br_chan->state != BT_L2CAP_DISCONNECTED) {
+			LOG_WRN("%s()%d: invalid transition", func, line);
```

Comment:
It's possible to configure logging so that function & line information is automatically included, i.e. it's redundant to do this as part of the format string itself.

Reply:
Ah, I see you're doing a similar trick here as the net_buf API, to catch the place where the higher level function was called, rather than the LOG_().